nixnet/database/database.py

```diff
@@ -29,14 +28,23 @@ class Database(object):
         database_name(str): The database alias or file pathname to open.
     """
     def __init__(self, database_name):
-        # type: (typing.Text) -> None
-        self._handle = None  # To satisfy `__del__` in case nxdb_open_database throws
-        self._handle = _funcs.nxdb_open_database(database_name)
+        # type: (typing.Union[int, typing.Text]) -> None
+        self.db_needs_closing = False  # To satisfy `__del__` in case nxdb_open_database throws
+
+        if isinstance(database_name, int):
+            # overload the 'database_name' parameter so the Cluster.database property
```

Comment:
This looks like an abuse of the parameter.  What if a user passes in a number?

One option is to have a keyword-only parameter thats name starts with "_".

Reply:
Yeah, I wasn't fond of this. I was going to post here and ask for advice but you beat me to it. I wish I could overload the constructor. Instead, I've overloaded the constructor parameter.

With this implementation, if the user passed in a number, the user would get an exception with NX_ERR_DATABASE_BAD_REFERENCE from the driver as soon as they tried to do anything with the object. That seems reasonable to me.

Otherwise, here are my thoughts:

1. I think I would prefer this functionality be private. The user has no reason to constructor a Database from a ref. It's only done internally for the `Cluster.database` back reference.
2. Question: Should the user expect to have to call `Database.close` on this object? I wouldn't think so, since they may just be doing something like `print(Cluster.database.name)`.

I'll look into the keyword-only parameter idea; it looks promising.

Reply:
> Question: Should the user expect to have to call Database.close on this object? I wouldn't think so, since they may just be doing something like print(Cluster.database.name).

This is multi-part
- From the user's perspective, its all the same database reference and I think they shouldn't have to close it but ...
- What do we need to do to satisfy the driver (whether it ref counts or not, the use of "close all refs", etc)?

Reply:
The driver keeps a count where nxdbOpenDatabase increments and nxdbCloseDatabase decrements (unless "close all refs" is true in which case the count goes to 0 and the db is closed).

So we just need to avoid calling nxdbOpenDatabase if we don't expect a matching call to nxdbCloseDatabase (with "close all refs" set to false), which is what I've done here.

Reply:
Well, I just realized one potential flaw. If you call `database.close()` twice on the same object, we emit a warning and return early the second time. Now, they can split the calls as follows:

db = Database(some_path)
cluster = db.clusters[cluster_name]
cluster.database.close()
db.close()

Reply:
I looked into keyword-only parameters but that seems to be only py3 with no backport to py2. I could still use keyword parameters (something like `def __init__(self, database_name, **_kwargs):` but I'm still not very fond of that, plus it still leaves us with the `Database.close()` issue I mentioned above.

I looked into having the Cluster constructor take in an existing Database object so it could just pass that back when `Cluster.database` is called. I had Database pass in a private factory method for Cluster into DbCollection that initializes the Cluster with the Database's self, but then realized properties of other classes (like `Ecu.cluster`) would also need to know about the Database object when they create a Cluster object.

Another option that I'm not too fond of; what if the Database constructor adds self to a private static list in Database, and Database exposes a private static method to get the Database object that is associated with a given handle? I can imagine a few pitfalls with this approach.

Still researching but I would welcome thoughts on this.

Reply:
RE private params

What I'd expect is a `**kwargs` and then we do `kwargs.get("_db_id")`.  nifpga has [an example of this](https://github.com/ni/nifpga-python/blob/master/nifpga/session.py#L95) though I have no idea why they iterate for the field.

RE `Cluster.database`

So let's reframe the situation.  I can understand doing a `db.find` to get a subframe and then calling `subframe.frame`.  Is there any context where the `Cluster.database` is required?

Reply:
I reverted Cluster.database for the moment. Perhaps I'll create an issue for it and we can address it later, and not as part of this huge review.

Reply:
Created issue #244